File: DebateUser/DebateUserFrontEnd/src/services/authService.js
/**
 * 인증 서비스 (Authentication Service)
 * 
 * 사용자 인증 관련 API 호출을 담당하는 서비스입니다.
 * 
 * 주요 기능:
 * - 로그인
 * - 회원가입
 * - 현재 사용자 정보 조회
 */

import api from './api'

export const authService = {
  /**
   * 로그인
   * 
   * 이메일과 비밀번호로 로그인합니다.
   * 
   * @param {string} email - 이메일
   * @param {string} password - 비밀번호
   * @returns {Promise<Object>} ApiResponse 구조의 응답 데이터
   * @returns {Object} response - ApiResponse { success: boolean, message: string, data: AuthResponse }
   * @returns {Object} response.data - AuthResponse { token: string, type: string, user: UserResponse }
   */
  async login(email, password) {
    // 인터셉터가 이미 ApiResponse 구조를 반환하므로 그대로 사용
    const response = await api.post('/auth/login', {
      email,
      password,
    })
    // ApiResponse 구조: { success: boolean, message: string, data: AuthResponse }
    // 인터셉터에서 이미 ApiResponse를 반환하므로 response가 ApiResponse입니다
    return response
  },

  /**
   * 회원가입
   * 
   * 새로운 사용자를 등록합니다.
   * 
   * @param {Object} registerData - 회원가입 데이터
   * @param {string} registerData.email - 이메일
   * @param {string} registerData.password - 비밀번호
   * @param {string} registerData.nickname - 닉네임
   * @returns {Promise<Object>} ApiResponse 구조의 응답 데이터
   * @returns {Object} response - ApiResponse { success: boolean, message: string, data: AuthResponse }
   * @returns {Object} response.data - AuthResponse { token: string, type: string, user: UserResponse }
   */
  async register(registerData) {
    // 인터셉터가 이미 ApiResponse 구조를 반환하므로 그대로 사용
    const response = await api.post('/auth/register', registerData)
    // ApiResponse 구조: { success: boolean, message: string, data: AuthResponse }
    // 인터셉터에서 이미 ApiResponse를 반환하므로 response가 ApiResponse입니다
    return response
  },

  /**
   * 현재 로그인한 사용자 정보 조회
   * 
   * JWT 토큰을 기반으로 현재 사용자의 정보를 가져옵니다.
   * 
   * @returns {Promise<Object>} ApiResponse 구조의 응답 데이터
   * @returns {Object} response - ApiResponse { success: boolean, message: string, data: UserResponse }
   * @returns {Object} response.data - UserResponse (사용자 정보)
   */
  async getCurrentUser() {
    // 인터셉터가 이미 ApiResponse 구조를 반환하므로 그대로 사용
    const response = await api.get('/users/me')
    // ApiResponse 구조: { success: boolean, message: string, data: UserResponse }
    // 인터셉터에서 이미 ApiResponse를 반환하므로 response가 ApiResponse입니다
    return response
  },
}

File: DebateUser/DebateUserBackEnd/src/main/java/com/debate/service/AuthService.java
package com.debate.service;

import com.debate.dto.request.LoginRequest;
import com.debate.dto.request.RegisterRequest;
import com.debate.dto.response.AuthResponse;
import com.debate.dto.response.UserResponse;
import com.debate.entity.User;
import com.debate.exception.BadRequestException;
import com.debate.exception.UnauthorizedException;
import com.debate.repository.UserRepository;
import com.debate.util.JwtUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * 인증(Authentication) 관련 비즈니스 로직을 처리하는 서비스 클래스
 * 회원가입, 로그인, JWT 토큰 생성 등의 기능을 제공합니다.
 */
@Service
@RequiredArgsConstructor
public class AuthService {
    private final UserRepository userRepository;      // 사용자 데이터 접근 리포지토리
    private final PasswordEncoder passwordEncoder;     // 비밀번호 암호화 인코더
    private final JwtUtil jwtUtil;                     // JWT 토큰 유틸리티

    /**
     * 회원가입 처리
     * 
     * @param request 회원가입 요청 데이터
     * @return 인증 응답 (JWT 토큰 및 사용자 정보)
     * @throws BadRequestException 이메일 또는 아이디가 이미 사용 중인 경우
     */
    @Transactional
    public AuthResponse register(RegisterRequest request) {
        // 이메일 중복 확인
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new BadRequestException("이미 사용 중인 이메일입니다");
        }


        // 사용자 엔티티 생성
        User user = User.builder()
                .email(request.getEmail())                                    // 이메일
                .password(passwordEncoder.encode(request.getPassword()))      // 비밀번호 암호화
                .nickname(request.getNickname())                              // 닉네임
                .bio(request.getBio())                                        // 자기소개
                .status(User.UserStatus.ACTIVE)                               // 상태: 활성
                .emailVerified(false)                                         // 이메일 인증: 미인증
                .build();

        // 사용자 저장
        user = userRepository.save(user);

        // JWT 토큰 생성
        String token = jwtUtil.generateToken(user.getId(), user.getEmail());

        // 인증 응답 생성
        return AuthResponse.builder()
                .token(token)                          // JWT 토큰
                .user(UserResponse.from(user))          // 사용자 정보
                .build();
    }

    // [추가] 이메일 중복 확인 (true면 중복)
    public boolean isEmailDuplicate(String email) {
        return userRepository.existsByEmail(email);
    }

    // [추가] 닉네임 중복 확인 (true면 중복)
    public boolean isNicknameDuplicate(String nickname) {
        return userRepository.existsByNickname(nickname);
    }

    /**
     * 로그인 처리
     * 이메일과 비밀번호로 인증합니다.
     *
     * @param request 로그인 요청 데이터
     * @return 인증 응답 (JWT 토큰 및 사용자 정보)
     * @throws UnauthorizedException 이메일 또는 비밀번호가 올바르지 않거나 계정이 비활성화된 경우
     */
    public AuthResponse login(LoginRequest request) {
        // 이메일로 사용자 조회
        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new UnauthorizedException("이메일 또는 비밀번호가 올바르지 않습니다"));

        // 비밀번호 확인
        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new UnauthorizedException("이메일 또는 비밀번호가 올바르지 않습니다");
        }

        // 계정 상태 확인 (활성화된 계정만 로그인 가능)
        if (user.getStatus() != User.UserStatus.ACTIVE) {
            throw new UnauthorizedException("비활성화된 계정입니다");
        }

        // JWT 토큰 생성
        String token = jwtUtil.generateToken(user.getId(), user.getEmail());

        // 인증 응답 생성
        return AuthResponse.builder()
                .token(token)                          // JWT 토큰
                .user(UserResponse.from(user))          // 사용자 정보
                .build();
    }
}

